kairos-prover/kairos-circuit-logic/src/transactions.rs

```diff
@@ -0,0 +1,541 @@
+                transaction: Transfer { recipient, amount },
+            };
+
+            match (
+                sender_balance.checked_sub(amount),
+                recipient_balance.checked_add(amount),
+            ) {
+                (Some(_), Some(_)) => {
+                    let sender_account = self.l2.accounts.get_mut(&sender).unwrap();
+                    sender_account.balance -= amount;
+                    sender_account.nonce += 1;
+
+                    self.l2.accounts.get_mut(&recipient).unwrap().balance += amount;
+
+                    (
+                        signed_transfer(sender, recipient),
+                        TxnExpectedResult::Success,
+                    )
+                }
+                _ => (
+                    signed_transfer(sender, recipient),
+                    TxnExpectedResult::Failure,
+                ),
+            }
+        }
+
+        pub fn random_withdraw(
+            &mut self,
+            sender: sample::Index,
+            amount: sample::Index,
+        ) -> (Signed<Withdraw>, TxnExpectedResult) {
+            let sender = self.l2.sample_keys(sender);
+
+            let sender_account = self
+                .l2
+                .accounts
+                .get(&sender)
+                .expect("sender does not have an l2 account in AccountsState");
+            let sender_balance = sender_account.balance;
+            let nonce = sender_account.nonce;
+
+            let l1_balance = self.l1.get_mut_or_insert_with(sender.clone(), || 0);
+
+            let amount = if sender_balance == 0 {
+                return (
+                    Signed {
+                        public_key: sender.deref().clone(),
+                        nonce,
+                        transaction: Withdraw { amount: 0 },
+                    },
+                    TxnExpectedResult::Failure,
+                );
+            } else {
+                amount.index(sender_balance as usize) as u64 + 1
+            };
+
+            let signed_withdraw = |public_key: PublicKey| Signed {
+                public_key,
+                nonce,
+                transaction: Withdraw { amount },
+            };
+
+            match (
+                sender_balance.checked_sub(amount),
+                l1_balance.checked_add(amount),
+            ) {
+                (Some(new_sender_bal), Some(new_recipient_bal)) => {
+                    let sender_account = self.l2.accounts.get_mut(&sender).unwrap();
+                    sender_account.balance = new_sender_bal;
+                    sender_account.nonce += 1;
+
+                    *l1_balance = new_recipient_bal;
+
+                    (signed_withdraw(sender.to_vec()), TxnExpectedResult::Success)
+                }
+                _ => (signed_withdraw(sender.to_vec()), TxnExpectedResult::Failure),
+            }
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct Accounts<A: Debug> {
+        pub pub_keys: Vec<Rc<PublicKey>>,
```

Comment:
we sample keys from the vector. hashmaps can't have an efficient random sample function. usize -> PublicKey

Reply:
Random samples are part of property testing.
In proptest or quickcheck you write code to generate pseudo random instances of your type and then test your code works as expected.

This is better than fixed testing because you cover more of your functions domain.

If your code fails the your test will be run again with a semantically smaller input. This is called shrinking. It allows you to find a minimal example that breaks your assumptions.